examples/lighting-app/linux/OnOffCluster.cpp

```diff
@@ -0,0 +1,38 @@
+
+#include <app/CommandHandler.h>
+#include <app/CommandSender.h>
+#include <app/MessageDef.h>
+#include <app/im-encoder.h>
+#include <app/util/basic-types.h>
+
+#include <support/logging/CHIPLogging.h>
+
+#include "LightingManager.h"
+
+namespace chip {
+namespace app {
+namespace cluster {
+namespace OnOff {
+
+void HandleOffCommandReceived(chip::TLV::TLVReader & aReader, chip::app::Command * apCommandObj)
```

Comment:
As I said, the fact that `examples/lighting-app` are the direct consumer of the IM looks wrong. 

If that was me, an in order to move forward with the encoder/decoder part of the IM, I will get rid of those callbacks, and get `DataModelHandler` to directly consume `chip::app::Command` so it can be dispatched properly to ZCL.
As a result it means there should be no changes to any of the `examples/` since IM will be considered an encoding/decoding scheme.

Reply:
I believe this file is quite misleading, so I moved it to src/app/clusters/on-off-server. Sorry for misleading.
I agree that the app should not handle commands directly, and there are some issues if we re-encode the message for current DataModelHandler.
The first one is the content of the message, as you can see, DataModelHandler takes raw packet from SecureSessionManager, but the Commands takes cooked message from ExchangeManager. There are some gaps between the SecureSessionManager and ExchangeManager.
Another issue is how to deal with outgoing messages. The content of return message by current zcl library is not consistent from the messages required by IM model.
After reading the code, I guess one possible solution is to call callbacks from src/app/clusters, (the function body is not implemented as this is not the target of this PR).

Reply:
About differences between incoming/outgoing messages from `SecureSessionManager`/`ExchangeManager` that's the part we need to figure out.

In fact this is where the hard part of the patch lives :)

Especially for outgoing messages I would say since the IM design possibly required the results of various commands and I believe this is not how the ZCL code is designed

As a stopgap I would first focus on dispatching an IM packet with a single command, that expects a single command response, and see how to integrate it into the current ZCL code. Once we are here it will requires a few other patches to get to the point where we can aggregate the result of various commands to be sent back as a response

Reply:
You can find that the interaction model engine is using std::map, I will submit another PR after this to remove the std::map from IME, and there might be a new API like `DispatchCommand(chip::ClusterId aClusterId, chip::CommandId aCommandId, chip::TLV::TLVReader & aReader, Command * apCommandObj, Command::CommandRoleId aCommandRoleId)`. Which should be useful for you to make tests.